rusoto/core/src/v2.rs

```diff
@@ -0,0 +1,79 @@
+//! TODO
+
+use futures::{IntoFuture, Future};
+
+use crate::Client;
+use crate::Region;
+use crate::{RusotoResult, RusotoFuture};
+use crate::error::RusotoError;
+use crate::request::HttpResponse;
+use crate::signature::SignedRequest;
+
+/// TODO
+pub trait ServiceRequest {
+    /// TODO
+    type Output;
+    /// TODO
+    type Error;
+
+    /// TODO
+    fn dispatch(self, region: &Region, dispatcher: &impl Dispatcher) -> RusotoFuture<Self::Output, Self::Error>;
+}
+
+/// TODO
+pub struct Request<T> {
+    input: T,
+    region: Region,
+    client: Client,
+}
+
+impl<T: ServiceRequest> Request<T> {
+    /// TODO
+    pub fn new(input: T, region: Region, client: Client) -> Self {
+        Request { input, region, client }
+    }
+
+    /// TODO
+    pub fn send(self) -> RusotoFuture<T::Output, T::Error> {
+        self.input.dispatch(&self.region, &self.client)
+    }
+
+    /// TODO
+    pub fn sync(self) -> RusotoResult<T::Output, T::Error>
+    where
+        T::Output: Send + 'static,
+        T::Error: Send + 'static,
+    {
+        self.send().sync()
+    }
+}
+
+impl<T: ServiceRequest> IntoFuture for Request<T> {
+    type Future = RusotoFuture<T::Output, T::Error>;
+    type Item = T::Output;
+    type Error = RusotoError<T::Error>;
+
+    fn into_future(self) -> Self::Future {
+        self.send()
+    }
+}
+
+/// TODO
+pub trait Dispatcher {
+/// TODO
+    fn dispatch<T, E>(
+        &self,
+        request: SignedRequest,
+        response_handler: fn(HttpResponse) -> Box<Future<Item = T, Error = RusotoError<E>> + Send>,
```

Comment:
Small thing & perhaps out of scope, but I wonder how feasible it'd be to replace the [`HttpResponse`](http://rusoto.github.io/rusoto/rusoto_core/request/struct.HttpResponse.html) with an `http:Response<hyper::Body>`? My motivation is to create a more modular Rusoto Core.

Reply:
I want to be careful to keep any `hyper` types out of the public API, but ultimately this could be something like `http::Resonse<some  body type>`.

Perhaps it would even be possible to use `tower_web::HttpService` instead of `Dispatcher` here, but we mightt want to wait until the tower ecosystem stabilises a little before that.

Reply:
> I want to be careful to keep any hyper types out of the public API, but yeah, ultimately this could be something like http::Resonse<some body type>.

That's a good call. If I were shorten my comment/question, I think I'd prefer to rely on types from `http` as much as possible (e.g., replace `SignedRequest` with `http::Request` and a struct that implements the `Header` trait from https://docs.rs/headers/0.2.1/headers/trait.Header.html. I'll see if I can send a pull request that does that.

> Perhaps it would even be possible to use `tower_web::HttpService` instead of Dispatcher here, but we mightt want to wait until the tower ecosystem stabilises a little before that.

Tower's `Service` can work! I wanted to use Tower's `Service` trait—or something that looks a lot like it—to handle retries, timeouts, and SigV4 signing, resulting a in a service layering that looks similar to `Timeout<Retry<SigV4<hyper::Client>>>` (for more on layering, see: https://docs.rs/tower/0.1.0/tower/builder/struct.ServiceBuilder.html#order). My biggest concern is that Tower doesn't support `std::future::Future` at the moment, which makes me think that it might be best to keep things as-is. There's only so much time in the day, after all.

Reply:
> keep any hyper types out of the public API

would putting `tower_web::HttpService` it its place put tower types in in the public API?

Reply:
at this point I would consider hyper types to be the least common denominator and winner of the rust crate ecosystem "best survivor" award in the category of http crates

Reply:
If I'm to prioritize what external crates are candidates for being placed in the public interface, `http` would win. My _slight_ preference for Tower over Hyper is that Tower's `Service` trait is essentially:

```rust
async fn call<T>(req: Request<T>) -> Result<Response<T>, Error>>
```

e.g., the surface area of Tower is significantly smaller than Hyper, but it does have the downside of being less mature and widespread than Hyper (hence, my preference for prioritizing exposing `http` types).

Reply:
I meant in terms of project longevity not so much in expressiveness or conciseness. A number of projects have tried to replace hyper have come and gone and another of projects that wrapped hyper have come and gone but hyper is the only crate that ended up standing the test of time.  It would be unfortunate to have rusoto to expose third party types of a crate that is exciting and new today but may not stand the test of time as well has hyper.